File: vision_fullcam/geometry/height.py
import numpy as np

def estimate_person_height_px(keypoints: dict) -> float | None:
    """
    keypoints: MoveNet output
    """
    try:
        ankle_y = np.mean([
            keypoints["left_ankle"][1],
            keypoints["right_ankle"][1],
        ])
        head_y = keypoints["nose"][1]
        return abs(ankle_y - head_y)
    except KeyError:
        return None

def estimate_ladder_height_m(
    ladder_bbox,
    person_height_px: float,
    assumed_person_height_m: float = 1.7,
) -> float | None:
    if person_height_px is None or person_height_px < 30:
        return None

    x1, y1, x2, y2 = ladder_bbox
    ladder_px = abs(y2 - y1)

    scale = assumed_person_height_m / person_height_px
    return ladder_px * scale

File: frontend/mobile/src/vision/geometry/height.ts
import type { BBox, Keypoint } from '../types';

const ASSUMED_PERSON_HEIGHT_M = 1.7;

export function estimatePersonHeightPx(
  keypoints: Record<string, Keypoint>,
): number | null {
  const leftAnkle = keypoints.left_ankle;
  const rightAnkle = keypoints.right_ankle;
  const nose = keypoints.nose;

  if (!leftAnkle || !rightAnkle || !nose) return null;
  if (leftAnkle.score < 0.3 || rightAnkle.score < 0.3 || nose.score < 0.3) {
    return null;
  }

  const ankleY = (leftAnkle.y + rightAnkle.y) / 2;
  const headY = nose.y;
  return Math.abs(ankleY - headY);
}

export function estimateLadderHeightM(
  ladderBbox: BBox,
  personHeightPx: number | null,
): number | null {
  if (personHeightPx === null || personHeightPx < 30) return null;

  const ladderPx = Math.abs(ladderBbox[3] - ladderBbox[1]);
  const scale = ASSUMED_PERSON_HEIGHT_M / personHeightPx;
  return ladderPx * scale;
}
